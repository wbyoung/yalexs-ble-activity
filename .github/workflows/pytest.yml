# yamllint disable-line rule:line-length
# see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-python

name: Pytest

on:
  push:
    branches: [main]
  pull_request:
    branches: [main]
  workflow_dispatch:
    inputs:
      upgraded:
        description: "Include upgraded dependencies in build matrix"
        default: false
        type: boolean
  schedule:
    - cron: "0 0 * * *"

jobs:
  info:
    name: Collect info
    outputs:
      dependencies: ${{ steps.info.outputs.dependencies }}
    runs-on: ubuntu-latest
    steps:
      - name: Collect info
        id: info
        run: |
          dependencies='["pinned", "upgraded"]'

          if [[ "${{ github.event_name }}" == "push" ]] || \
             [[ "${{ github.event_name }}" == "pull_request" ]] || \
             [[ "${{ github.event.inputs.upgraded }}" == "false" ]];
          then
            dependencies='["pinned"]'
          fi

          # Output & sent to GitHub Actions
          echo "dependencies: ${dependencies}"
          echo "dependencies=${dependencies}" >> $GITHUB_OUTPUT
  build:
    name: Run tests
    runs-on: ubuntu-latest
    needs:
      - info
    strategy:
      fail-fast: false
      matrix:
        dependencies: ${{ fromJSON(needs.info.outputs.dependencies) }}
        python-version: ["3.13"]

    steps:
      - uses: actions/checkout@v4
      - name: Set up Python ${{ matrix.python-version }}
        uses: actions/setup-python@v5
        with:
          python-version: ${{ matrix.python-version }}
      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip

          if [[ "${{ matrix.dependencies }}" == "upgraded" ]];
          then
            echo "unpinning dependencies"
            ./script/unfreeze
          fi

          pip install -r requirements.test.txt

          if [[ "${{ matrix.dependencies }}" == "upgraded" ]];
          then
            ./script/freeze
            echo "dependency changes after upgrade:"
            git diff requirements.test.txt
          fi
      - name: Test with pytest
        run: |
          pytest -p no:sugar
      - name: Validate coverage
        run: |
          coverage json -q -o - | python3 .github/validate_coverage.py
