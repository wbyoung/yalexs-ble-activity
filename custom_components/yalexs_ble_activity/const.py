"""Constants for the Yale Access Bluetooth Activity integration."""

from typing import Final

DOMAIN: Final = "yalexs_ble_activity"
YALEXSBLE_PATCH_URL = (
    "git+https://github.com/wbyoung/yalexs-ble@yalexs-ble-{version}-patches"
)

ATTR_REMOTE_TYPE: Final = "remote_type"
ATTR_SLOT: Final = "slot"
ATTR_SOURCE: Final = "source"
ATTR_TIMESTAMP: Final = "timestamp"

CONF_LOCK_ENTITIES: Final = "lock_entities"

OPERATION_SENSOR_WRITE_DELAY: Final = 2

TRACE: Final = 5
