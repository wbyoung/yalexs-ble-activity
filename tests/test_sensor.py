"""Test Yale Access Bluetooth Activity sensors."""

from typing import Any
from unittest.mock import patch

from homeassistant.const import Platform
from homeassistant.core import HomeAssistant, State
from homeassistant.helpers import entity_registry as er
from homeassistant.helpers.restore_state import STORAGE_KEY as RESTORE_STATE_KEY
import pytest
from pytest_homeassistant_custom_component.common import (
    MockConfigEntry,
    async_mock_restore_state_shutdown_restart,
    mock_restore_cache_with_extra_data,
    snapshot_platform,
)
from syrupy.assertion import SnapshotAssertion
from yalexs_ble import DoorActivity, LockActivity
from yalexs_ble.const import (
    DoorStatus,
    LockOperationRemoteType,
    LockOperationSource,
    LockStatus,
)

from . import MOCK_UTC_NOW, MockNow, setup_integration


async def test_sensors(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
    lock: er.RegistryEntry,
    entity_registry: er.EntityRegistry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test all sensors created by the integration."""
    with patch("custom_components.yalexs_ble_activity.PLATFORMS", [Platform.SENSOR]):
        await setup_integration(hass, config_entry)

    await snapshot_platform(hass, entity_registry, snapshot, config_entry.entry_id)


@pytest.mark.parametrize(
    "lock_activities",
    [
        [
            LockActivity(
                timestamp=MOCK_UTC_NOW,
                status=LockStatus.UNLOCKED,
                source=LockOperationSource.PIN,
                remote_type=LockOperationRemoteType.UNKNOWN,
                slot=3,
            ),
            LockActivity(
                timestamp=MOCK_UTC_NOW,
                status=LockStatus.LOCKED,
                source=LockOperationSource.AUTO_LOCK,
                remote_type=None,
                slot=None,
            ),
        ],
        [
            LockActivity(
                timestamp=MOCK_UTC_NOW,
                status=LockStatus.LOCKED,
                source=LockOperationSource.REMOTE,
                remote_type=LockOperationRemoteType.BLE,
                slot=None,
            )
        ],
        [
            DoorActivity(
                timestamp=MOCK_UTC_NOW,
                status=DoorStatus.AJAR,
            )
        ],
        [type("UnsupportedActivity", (object,), {"timestamp": MOCK_UTC_NOW})()],
    ],
    ids=[
        "pin_unlock_then_auto_lock",
        "remote_lock",
        "door_ajar",
        "unsupported_activity",
    ],
)
async def test_sensor_activity_update(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
    lock: er.RegistryEntry,
    lock_activities: LockActivity,
    now: MockNow,
    entity_registry: er.EntityRegistry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test all sensors created by the integration."""
    with patch("custom_components.yalexs_ble_activity.PLATFORMS", [Platform.SENSOR]):
        await setup_integration(hass, config_entry)

    def _snapshot(phase: str):
        for entity_entry in er.async_entries_for_config_entry(
            entity_registry, config_entry.entry_id
        ):
            assert entity_entry == snapshot(
                name=f"{entity_entry.entity_id}-{phase}-entry"
            )
            assert entity_entry.disabled_by is None, "Please enable all entities."
            state = hass.states.get(entity_entry.entity_id)
            assert state, f"State not found for {entity_entry.entity_id}"
            assert state == snapshot(name=f"{entity_entry.entity_id}-{phase}-state")

    activity_update = _activity_update_handler(hass, lock)

    for lock_activity in lock_activities:
        activity_update(lock_activity, lock_info=None, connection_info=None)

    _snapshot("activity-received")

    now._tick(2)
    await hass.async_block_till_done()
    _snapshot("post-tick")


RESTORE_STATE_PARAMETRIZED = ("stored_data", "expected_state", "expected_attributes")
RESTORE_STATE_SCENARIOS = {
    "simple": {
        "stored_data": {
            "value": "locked",
            "attributes": {
                "source": "auto_lock",
            },
        },
        "expected_state": "locked",
        "expected_attributes": {
            "source": "auto_lock",
        },
    },
    "no_values": {
        "stored_data": {"value": None, "attributes": None},
        "expected_state": "unknown",
        "expected_attributes": {},
    },
}


@pytest.mark.parametrize(
    RESTORE_STATE_PARAMETRIZED,
    [
        tuple(scenario.get(param) for param in RESTORE_STATE_PARAMETRIZED)
        for scenario in RESTORE_STATE_SCENARIOS.values()
    ],
    ids=RESTORE_STATE_SCENARIOS.keys(),
)
async def test_restore_sensor_save_state(
    hass: HomeAssistant,
    hass_storage: dict[str, Any],
    config_entry: MockConfigEntry,
    lock: er.RegistryEntry,
    stored_data: dict[str, Any],
    expected_state: str,
    expected_attributes: dict[str, Any],
    entity_registry: er.EntityRegistry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test saving sensor/orchestrator state."""
    await setup_integration(hass, config_entry)

    entity_id = "sensor.front_door_operation"
    operation_entity = _activity_update_handler(hass, lock).__self__
    operation_entity._attr_native_value = (
        expected_state if expected_state != "unknown" else None
    )
    operation_entity._attr_extra_state_attributes = expected_attributes or None

    await hass.async_block_till_done()
    await async_mock_restore_state_shutdown_restart(hass)  # trigger saving state

    stored_entity_data = [
        item["extra_data"]
        for item in hass_storage[RESTORE_STATE_KEY]["data"]
        if item["state"]["entity_id"] == entity_id
    ]

    assert stored_entity_data[0] == stored_data
    assert stored_entity_data == snapshot


@pytest.mark.parametrize(
    RESTORE_STATE_PARAMETRIZED,
    [
        tuple(scenario.get(param) for param in RESTORE_STATE_PARAMETRIZED)
        for scenario in RESTORE_STATE_SCENARIOS.values()
    ],
    ids=RESTORE_STATE_SCENARIOS.keys(),
)
async def test_restore_state(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
    lock: er.RegistryEntry,
    stored_data: dict[str, Any],
    expected_state: str,
    expected_attributes: dict[str, Any],
    snapshot: SnapshotAssertion,
) -> None:
    """Test restoring sensor/orchestrator state."""
    entity_id = "sensor.front_door_operation"
    mock_restore_cache_with_extra_data(
        hass,
        (
            (
                State(
                    entity_id,
                    "mock-state",  # note: in reality, this would match the stored data
                ),
                stored_data,
            ),
        ),
    )

    await setup_integration(hass, config_entry)

    state = hass.states.get(entity_id)
    assert state
    assert state.state == expected_state
    assert {
        key: value
        for key, value in state.attributes.items()
        if key not in {"friendly_name", "icon"}
    } == expected_attributes


def _activity_update_handler(hass: HomeAssistant, lock: er.RegistryEntry):
    core_entry = hass.config_entries.async_get_known_entry(lock.config_entry_id)
    data = core_entry.runtime_data
    register_activity_update_call = data.lock.register_activity_callback.mock_calls[-1]
    _name, register_activity_update_args, _kwargs = register_activity_update_call
    (activity_update,) = register_activity_update_args

    return activity_update
