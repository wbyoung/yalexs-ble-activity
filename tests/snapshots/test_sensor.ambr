# serializer version: 1
# name: test_restore_sensor_save_state[no_values]
  list([
    dict({
      'attributes': None,
      'value': None,
    }),
  ])
# ---
# name: test_restore_sensor_save_state[simple]
  list([
    dict({
      'attributes': dict({
        'source': 'auto_lock',
      }),
      'value': 'locked',
    }),
  ])
# ---
# name: test_sensor_activity_update[door_ajar][sensor.front_door_operation-activity-received-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.front_door_operation',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:lock-clock',
    'original_name': 'Operation',
    'platform': 'yalexs_ble_activity',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'operation',
    'unique_id': 'mock-address:front_dooroperation',
    'unit_of_measurement': None,
  })
# ---
# name: test_sensor_activity_update[door_ajar][sensor.front_door_operation-activity-received-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Front door Operation',
      'icon': 'mdi:lock-clock',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.front_door_operation',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unknown',
  })
# ---
# name: test_sensor_activity_update[door_ajar][sensor.front_door_operation-post-tick-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.front_door_operation',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:lock-clock',
    'original_name': 'Operation',
    'platform': 'yalexs_ble_activity',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'operation',
    'unique_id': 'mock-address:front_dooroperation',
    'unit_of_measurement': None,
  })
# ---
# name: test_sensor_activity_update[door_ajar][sensor.front_door_operation-post-tick-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Front door Operation',
      'icon': 'mdi:lock-clock',
      'timestamp': datetime.datetime(2025, 5, 20, 10, 51, 32, 3245, tzinfo=datetime.timezone.utc),
    }),
    'context': <ANY>,
    'entity_id': 'sensor.front_door_operation',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'door_ajar',
  })
# ---
# name: test_sensor_activity_update[pin_unlock_then_auto_lock][sensor.front_door_operation-activity-received-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.front_door_operation',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:lock-clock',
    'original_name': 'Operation',
    'platform': 'yalexs_ble_activity',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'operation',
    'unique_id': 'mock-address:front_dooroperation',
    'unit_of_measurement': None,
  })
# ---
# name: test_sensor_activity_update[pin_unlock_then_auto_lock][sensor.front_door_operation-activity-received-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Front door Operation',
      'icon': 'mdi:lock-clock',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.front_door_operation',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unknown',
  })
# ---
# name: test_sensor_activity_update[pin_unlock_then_auto_lock][sensor.front_door_operation-post-tick-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.front_door_operation',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:lock-clock',
    'original_name': 'Operation',
    'platform': 'yalexs_ble_activity',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'operation',
    'unique_id': 'mock-address:front_dooroperation',
    'unit_of_measurement': None,
  })
# ---
# name: test_sensor_activity_update[pin_unlock_then_auto_lock][sensor.front_door_operation-post-tick-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Front door Operation',
      'icon': 'mdi:lock-clock',
      'source': 'auto_lock',
      'timestamp': datetime.datetime(2025, 5, 20, 10, 51, 32, 3245, tzinfo=datetime.timezone.utc),
    }),
    'context': <ANY>,
    'entity_id': 'sensor.front_door_operation',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'lock_locked',
  })
# ---
# name: test_sensor_activity_update[remote_lock][sensor.front_door_operation-activity-received-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.front_door_operation',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:lock-clock',
    'original_name': 'Operation',
    'platform': 'yalexs_ble_activity',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'operation',
    'unique_id': 'mock-address:front_dooroperation',
    'unit_of_measurement': None,
  })
# ---
# name: test_sensor_activity_update[remote_lock][sensor.front_door_operation-activity-received-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Front door Operation',
      'icon': 'mdi:lock-clock',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.front_door_operation',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unknown',
  })
# ---
# name: test_sensor_activity_update[remote_lock][sensor.front_door_operation-post-tick-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.front_door_operation',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:lock-clock',
    'original_name': 'Operation',
    'platform': 'yalexs_ble_activity',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'operation',
    'unique_id': 'mock-address:front_dooroperation',
    'unit_of_measurement': None,
  })
# ---
# name: test_sensor_activity_update[remote_lock][sensor.front_door_operation-post-tick-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Front door Operation',
      'icon': 'mdi:lock-clock',
      'remote_type': 'ble',
      'source': 'remote',
      'timestamp': datetime.datetime(2025, 5, 20, 10, 51, 32, 3245, tzinfo=datetime.timezone.utc),
    }),
    'context': <ANY>,
    'entity_id': 'sensor.front_door_operation',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'lock_locked',
  })
# ---
# name: test_sensor_activity_update[unsupported_activity][sensor.front_door_operation-activity-received-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.front_door_operation',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:lock-clock',
    'original_name': 'Operation',
    'platform': 'yalexs_ble_activity',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'operation',
    'unique_id': 'mock-address:front_dooroperation',
    'unit_of_measurement': None,
  })
# ---
# name: test_sensor_activity_update[unsupported_activity][sensor.front_door_operation-activity-received-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Front door Operation',
      'icon': 'mdi:lock-clock',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.front_door_operation',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unknown',
  })
# ---
# name: test_sensor_activity_update[unsupported_activity][sensor.front_door_operation-post-tick-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.front_door_operation',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:lock-clock',
    'original_name': 'Operation',
    'platform': 'yalexs_ble_activity',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'operation',
    'unique_id': 'mock-address:front_dooroperation',
    'unit_of_measurement': None,
  })
# ---
# name: test_sensor_activity_update[unsupported_activity][sensor.front_door_operation-post-tick-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Front door Operation',
      'icon': 'mdi:lock-clock',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.front_door_operation',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unknown',
  })
# ---
# name: test_sensors[sensor.front_door_operation-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.front_door_operation',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:lock-clock',
    'original_name': 'Operation',
    'platform': 'yalexs_ble_activity',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'operation',
    'unique_id': 'mock-address:front_dooroperation',
    'unit_of_measurement': None,
  })
# ---
# name: test_sensors[sensor.front_door_operation-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Front door Operation',
      'icon': 'mdi:lock-clock',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.front_door_operation',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unknown',
  })
# ---
